R/govcan_dl_resources.R

```diff
@@ -4,121 +4,147 @@
+                                              ...) {
+  fmt <- tolower(resources$format)
+  url <- resources$url
+
+  msgInfo(resources$name, paste0("(", fmt, ")"), "",
+    appendLF = FALSE)
+
+  if (grepl("^ftp://", url)) {
+    out <- empty_entry()
+    msgWarning("skipped (ftp not supported yet).")
   } else {
-    msgWarning("No match, no download!")
+    # select type of files to be downloaded
+    tmp <- TRUE
+    if (!is.null(excluded))
+      tmp <- !(fmt %in% tolower(excluded))
+    if (!is.null(included))
+      tmp <- fmt %in% tolower(included)
+
+    if (tmp) {
+      # extract filename from url 
+      fl <- extract_filename(url)
+      if (!is.null(fl)) {
+        if (id_as_filename) {
+          fl <- paste0(resources$id, ".", extract_extension(fl))
+          flp <- normalizePath(file.path(path, fl), mustWork = FALSE)
+        } else {
+          flp <- normalizePath(file.path(path, fl), mustWork = FALSE)
+        }
+        if (file.exists(flp)) {
+          # prevents from downloading the same file several times
+          msgWarning("skipped (already downloaded).")
+          out <- empty_entry("disk", fmt = fmt, path = flp)
+        } else {
+          out <- ckanr::ckan_fetch(url, format = fmt,
+              store = "disk", path = flp)
+          msgSuccess()
+        }
+      } else {
+        out <- empty_entry(fmt = fmt)
+        msgWarning("skipped (not supported).")
+      }
+    } else {
+      out <- empty_entry(fmt = fmt)
+      msgWarning("skipped (format not selected).")
+    }
   }
+
+  out$url <- url
+  out$package_id <- resources$package_id
+  out$id <- resources$id
+  out <- null_to_na(out)
+  out <- as.data.frame(out)
+  class(out) <- c("tbl_df", "tbl", "data.frame")
+  ord <- c("id", "package_id", "url", "path", "fmt")
+  out[, c(ord, setdiff(names(out), ord))]
 }
 
-#' @describeIn govcan_dl_resources Method for ckan_resource_stack objects.
+#' @describeIn govcan_dl_resources Method for `ckan_resource_stack` objects.
 #' @export
-govcan_dl_resources.ckan_resource_stack <- function(resources,
-                                                    file_formats = c("CSV"),
-                                                    where = getwd(), ...){
-
-  all_formats <- unlist(purrr::map(resources, ~.x$format))
-  wanted_indices <- which(all_formats %in% file_formats)
-
-  if (length(wanted_indices)){
-    if (where == "session"){
-
-      message("Warning: the session option is currently not working well due to issues in ckanr")
-
-      for (resource in wanted_indices){
-        resource_tmp <-  resources[[resource]]
-
-        ckanr::ckan_fetch(resource_tmp$url, store = "session")
-        write_import_message(resource_tmp)
-      }
-
-    } else if (where != "session"){
-
-      for (resource in wanted_indices){
-
-        resource_tmp <- resources[[resource]]
+govcan_dl_resources.ckan_resource_stack <- function(resources, ...) {
+  out <- lapply(resources, govcan_dl_resources, ...)
+  do.call(rbind, out)
```

Comment:
@KevCaz Can you explain the use of `do.call` for me here? I often see it but I'm a little fuzzy on the sue. Is it the way to apply rbind on a list? Why not use `bind_rows`?

Reply:
`do.call(ribind, list_of_df)` is just a relatively efficient to do row binding using only base R. The number of dependencies doesn't matter for the CRAN, however reviewers/editors at ropensci may raise a flag for unnecessary dependencies. I wouldn't be surprised if they raise a flag for `purrr` here, because all can be done with base R.